### PR TITLE
Adding tools.md

### LINKED ISSUE #144: Add contributing documentation to the MM repo.
- [x] readme.md - High level overview, building, and getting started.
   - [x] High level overview
   - [x] Preparing baserom
   - [x] How to set up the repo
   - [x] How to build the repo
   - [x] Day to day workflow (diff.py and maybe symbol rename)
- [x] contributing.md - Contributor's guidelines
   - [x] How to submit a PRs
   - [x] How to respond to reviewer comments
   - [x] Documentation rules
   - [x] Style and naming schemes
   - [x] Build warnings
   - [x] Updating trello cards
- [x] reviewing.md - What to do when reviewing code.
   - [x] When reviewing other people's code please follow the checklist.
   - [x] When someone addresses all of your comments you should approve the changes.
   - [x] If someone does not address your comments and expresses that a different way is better than yours, look for feedback from other contributors. The project lead will have final say in these situations, but the decisions will generally be guided by a consensus of contributors.
   - [x] Reviewer checklist. This is mostly the same as contributor submission checklist.
      - [x] `format.sh` was run.
      - [x] Files with NON_MATCHING functions are equivalent.
      - [x] `make` builds a matching ROM (jenkins)
      - [x] Any new compiler warnings that were added are required for matching (check the file that stores warnings)
      - [x] Comments and variables have correct spelling.
      - [x] `code` and `boot` segment files are documented. Overlays with documentation are complete.
         - [x] Overlays with documentation are required to have macros to define access to parameters if the parameter uses bitwise access. The params should have an enum otherwise.
      - [x] The following fields are declared in an Actor header file.
         - [x] Main Actor struct
         - [x] Extern'd initVar data.
         - [x] Types used in the actor struct. Specific example would be actionFunc typedefs.
         - [x] Param field macros and/or enums.
         - [x] We have proof that enum/struct/define is needed in another file.
      - [x] New variables & functions should follow standard naming conventions.
         - [x] Constants converted to whichever looks best in context: hexadecimal, decimal, or float
            - [x] Rotation angles should always be in hexadecimal
         - [x] Functions, structs, unions, enums, and typedefs are `TitleCase` (`DmRavine`)
            - [x] "Methods" for objects separate the system from the verb with an underscore (`DmRavine_Init`)
         - [x] Variable names are `camelCase` (`actionFunc`)
            - [x] Global variables start with `g` (`gSaveContext`)
            - [x] Static global variables start with `s` (`sSphereInit`)
         - [x] Macros and enum constants are `SCREAMING_SNAKE_CASE` (`DM_RAVINE_STATE_ACTIVE`)
         - [x] Trailing commas in array and struct definitions (see `EnJcMatoDamageTable`)
- [ ] tools.md - description and how to use for the various manually run tools in the repository (diff.py, regconvert, permuter, etc)
   - [ ] Regconvert
   - [ ] Gfxdis
   - [ ] How to add new warnings
   - [ ] Diff.py
   - [ ] Permuter
   - [ ] fixle.sh
   - [ ] format.sh
   - [ ] mips_to_c
   - [ ] gen_mips_to_c_context.py
   - [ ] ichaindis.py
   - [ ] colliderinit.py
   - [ ] actor_symbols.py
   - [ ] TODO: Any other tools need documentation?
- [ ] tutorial.md - table of contents for the decompilation tutorial.

### PR DESCRIPTION
**Pull request by jpburnett**
Before opening this PR, ensure the following:
- `./format.sh` was run to apply standard formatting.
- `make` successfully builds a matching ROM.
- No new compiler warnings were introduced during the build process.
    - Can be verified locally by running `tools/warnings_count/check_new_warnings.sh`
- New variables & functions should follow standard naming conventions.
- Comments and variables have correct spelling.
---

Getting back in to the swing of things I thought I would refresh myself on some of the tools. I filled in what I could / what I knew about the tools. This is also an effort to close #144 to clean things up. Feel free to let me know what is missing. This isn't a high priority to get it.

### CODE REVIEW

**Review by AngheloAlf — COMMENTED** [review]
Looks good overall.
I think extract_assets.py and graphovl.py are missing tho

**Review by Kenix3 — COMMENTED** [review]
This PR is also probably a good place to explain how to use the google sheet reservation board.

File: docs/tools.md [hunk 1]
```diff
@@ -0,0 +1,261 @@
...
+
+Formats all C files in the repo using `clang-format-11` (instructions on how to install this version are pinned in Discord if you can't get it from your package manager in the usual way). This will touch all files in the repo, so the next `make` will take longer.
+
+## External tools
+
+### mips_to_c
+
+mips_to_c (or mips2c) is a tool that takes MIPS assembly and will attempt to convert it to C.
+
+There are different ways you can run mpips2c:
```
[the remark this answers is not in the document]
Comment (jpburnett) [reply docs/tools.md:190]: Sounds good, I can add a section on that

File: docs/tools.md [hunk 2]
```diff
@@ -259,3 +278,7 @@ Extracts a damagetable from its address. Can also reconvert existing damagetable
 ### `tools/vtxdis`
 
 Extracts vertex data from a file. Essentially irrelevant since MM requires extracting this data with ZAPD anyway.
+
+### Reservation Tracking
+
+We use a [Google Sheet](https://docs.google.com/spreadsheets/d/1X83YCPRa532v-Zo0WgUsJ2kB1X9RxBta5_p9aWA8uro/edit#gid=0>) to keep track of all the files worked on and their merge status. On this board one can reserve a file to work on. It is common courtesy to not work on a file that is being worked on by another contributor. If there is a specific file you are interested in and it is reserved, reach out to the individual who has it reserved to come to some agreement.
```
Comment (EllipticEllipsis) [inline docs/tools.md:284]: Probably mention what "interested" means too.

File: docs/tools.md [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+There are different ways you can run mips2c:
+
+- Clone the repo and run it loclly on your own machine. Follow set up instructions at <https://github.com/matt-kempster/mips_to_c>
+- An online version is available at <https://simonsoftware.se/other/mips_to_c.py>.
+
+mips_to_c's accuracy can be improved when some context to what the existing C source code is like. See more on how to generate this in [m2ctx](#m2ctx.py)
+
+### Permuter
+
+(Exists in a subrepo.)
```
Comment (hensldm) [inline docs/tools.md:218]: This is no longer true, should just remove it.

**Review by hensldm — CHANGES_REQUESTED** [review]

**Review by AngheloAlf — APPROVED** [review]

File: docs/tools.md [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+Extracts vertex data from a file. Essentially irrelevant since MM requires extracting this data with ZAPD anyway.
+
+### Reservation Tracking
+
+We use a [Google Sheet](https://docs.google.com/spreadsheets/d/1X83YCPRa532v-Zo0WgUsJ2kB1X9RxBta5_p9aWA8uro/edit#gid=0>) to keep track of all the files worked on and their merge status. On this board one can reserve a file to work on. To reserve a file, put your github/discord name in the "Reserved" column. It is common courtesy to not work on a file that is being worked on by another contributor, so ensure the "Reserved" column is blank before working on a file. If you really want to work on a specific file and it is reserved you can put your name in the "Interested" column then reach out to the individual who has reserved the file and to see if they are still interested in pursuing the file. Other columns of interest can be the "Size" columns. The smaller the size, generally means it is easier to decompile. We recommend new comers to work on the smaller sized files(Roughly under the Average size of 40). After doing a couple we encourage you to move on to moderate sized files as to maintain smaller files for other new contributors.
```
Comment (EllipticEllipsis) [inline docs/tools.md:282]:
```suggestion
We use a [Google Sheet](https://docs.google.com/spreadsheets/d/1X83YCPRa532v-Zo0WgUsJ2kB1X9RxBta5_p9aWA8uro/edit#gid=0) to keep track of all the files' decompilation statuses and their function statistics. On this board one can reserve a file to work on. The columns are

- **Actor Id**: The enum that uniquely identifies the actor for spawning etc., useful for looking in KZ/VerboseOcarina/Spectrum. For `boot`/`code`, the VRAM of the file's .text start is given instead.
- **Overlay/File name**: Hopefully obvious, although some remain unknown.
- **Function size statistics**: Intended as a crude estimate of how hard a file will be. Beginners should look for small largest function size and total size; the columns give a rough estimate of the distribution of function sizes without getting unnecessarily statistically descriptive. As you become more experienced, you should work on larger files to 
- **Description**: What the file is. It's helpful if you can fill this in if you know! They should be synchronised with the short top-of-file descriptions.
- **Status**: (Free)/Reserved/PR/Merged. To be kept up-to-date by the reserver.
- **Reserved**: To reserve a file, put your Discord name in the "Reserved" column. It is common courtesy to not work on a file that is being worked on by another contributor, so ensure the "Reserved" column is blank before working on a file. If it is not, you can ask the reserver(s) if they want to release it or collaborate on it, but don't expect them to agree. More information on what is expected when you reserve a file is available in the [CONTRIBUTING.md](../CONTRIBUTING.md).
- **Interested**: If you would like to work on a file, but don't want to reserve it, or would be interested in collaboration, etc. You should talk to any Interested people if you want to work on the file.
- **Notes**: Any other useful information: partial progress by someone unable to finish the file, other files it works with, etc.
```

**Review by EllipticEllipsis — COMMENTED** [review]
I'm happy apart from the sheet description: I've rewritten the paragraph into some explicit column-describing bullet points, see what you think.

**Review by hensldm — APPROVED** [review]

File: docs/tools.md [hunk 5]
```diff
@@ -279,4 +279,13 @@ Extracts vertex data from a file. Essentially irrelevant since MM requires extra
...
+We use a [Google Sheet](https://docs.google.com/spreadsheets/d/1X83YCPRa532v-Zo0WgUsJ2kB1X9RxBta5_p9aWA8uro/edit#gid=0) to keep track of all the files' decompilation statuses and their function statistics. On this board one can reserve a file to work on. The columns are
+
+- **Actor Id**: The enum that uniquely identifies the actor for spawning etc., useful for looking in KZ/VerboseOcarina/Spectrum. For `boot`/`code`, the VRAM of the file's .text start is given instead.
+- **Overlay/File name**: Name of the Overlay or File in the codebase.
+- **Function size statistics**: Intended as a crude estimate of how hard a file will be. Beginners should look for small largest function size and total size; the columns give a rough estimate of the distribution of function sizes without getting unnecessarily statistically descriptive. As you become more experienced, you should work on larger files to
```
Comment (EllipticEllipsis) [inline docs/tools.md:286]:
```suggestion
- **Function size statistics**: Intended as a crude estimate of how hard a file will be. Beginners should look for small largest function size and total size; the columns give a rough estimate of the distribution of function sizes without getting unnecessarily statistically descriptive. As you become more experienced, you should work on larger files to leave the smaller ones for other beginners.
```

Ah, apparently this sentence didn't finish itself?

**Review by Kenix3 — APPROVED** [review]

**Pull request merged**